server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java

```diff
@@ -780,6 +780,25 @@ public static boolean isGlobalStateEquals(Metadata metadata1, Metadata metadata2
         return true;
     }
 
+    /**
+     * Reconciles the cluster state metadata taken at the end of a snapshot with the data streams and indices
+     * contained in the snapshot. Certain actions taken during a snapshot such as rolling over a data stream
+     * or deleting a backing index may result in situations where some reconciliation is required.
+     *
+     * @return Reconciled {@link Metadata} instance
+     */
+    public static Metadata snapshot(Metadata metadata, List<String> dataStreams, List<String> indices) {
+        var builder = Metadata.builder(metadata);
+        for (var dsName : dataStreams) {
+            var dataStream = metadata.dataStreams().get(dsName);
+            if (dataStream == null) {
+                throw new IllegalArgumentException("unable to find data stream [" + dsName + "]");
```

Comment:
Maybe add a comment here, that a data stream can't be deleted while snapshot is currently being made (that would include this data stream). So throwing an error is an accepted outcome here.